Hazelcast Java client not resolving member address when trying to reconnect

Context: basic HZ setup in a docker environment : 1 server + 1 client, HZ configuration based on tcp-ip with network aliases (using docker service names that are resolved against docker internal DNS).
When HZ server restarts, docker reassigns a new IP address to the container but the client forever tries to reconnect to the old IP address without re-resolving the member name to obtain the new IP.

**Version used :** 3.12.7 (also tested on 4.0.1 : same issue)
**Cluster size :** single member
**Number of clients :** 1 Java client
**Java version :** 1.8 on client side / 11 on server side (Note : DNS network ttl cache is set to 0 everywhere)
**OS :** Host: CentOS with docker engine (Docker version 19.03.6). Client container with debian based image. Server container from public HZ image: Alpine Linux v3.8.
**Logs :** see below in the steps

**Steps to reproduce :**
Using Docker (docker-compose) :
Create a HZ single member cluster using the standard public image, with tcp-ip member configured with the Docker service alias names (https://hub.docker.com/r/hazelcast/hazelcast/)
Server configuration example :
```
            <tcp-ip enabled="true">
              <member-list>
                <member>hazelcast:5701</member>
              </member-list>
            </tcp-ip>
```
("hazelcast" here is the service name in docker-compose for the HZ server, it's just a standard network alias that gets resolved on the docker network)

Connect a Java client configured with cluster-members filled with the same Docker service alias name.
Client configuration example :
```
    <network>
        <cluster-members>
            <address>hazelcast:5701</address>
        </cluster-members>
    </network>
```
On server startup, HZ successfully starts up. Logs shows the member configured as:
```
hazelcast     | Members {size:1, ver:1} [
hazelcast     |      Member [hazelcast]:5701 - ed12db96-27b4-4f05-b99b-840799a6bebf this
hazelcast     | ]
```
On client startup, HZ resolves the IP address of the server and connects to the cluster:
```
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to cluster with name: hazelcast
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.internal-3] com.hazelcast.client.connection.ClientConnectionManager.null hz.client_1 [hazelcast] [3.12.7] Setting ClientConnection{alive=true, connectionId=1, channel=NioChannel{/172.20.0.40:57827->hazelcast/172.20.0.39:5701}, remoteEndpoint=[hazelcast]:5701, lastReadTime=2020-06-08 12:25:44.343, lastWriteTime=2020-06-08 12:25:44.317, closedTime=never, connected server version=3.12.7} as owner with principal ClientPrincipal{uuid='cd0aefaa-df16-4699-88b5-a16e9f476c5d', ownerUuid='239fa3d4-8e13-4f3f-a11e-f3539db2a6fb'}
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.internal-3] com.hazelcast.client.connection.ClientConnectionManager.null hz.client_1 [hazelcast] [3.12.7] Authenticated with server [hazelcast]:5701, server version:3.12.7 Local address: /172.20.0.40:57827
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.event-5] com.hazelcast.client.spi.impl.ClientMembershipListener.null hz.client_1 [hazelcast] [3.12.7]
hz-test-client              |
hz-test-client              | Members [1] {
hz-test-client              |      Member [hazelcast]:5701 - 239fa3d4-8e13-4f3f-a11e-f3539db2a6fb
hz-test-client              | }
hz-test-client              |
hz-test-client              | 2020-06-08 12:25:44 INFO [hz.client_1.cluster-] com.hazelcast.core.LifecycleService.null hz.client_1 [hazelcast] [3.12.7] HazelcastClient 3.12.7 (20200430 - bf4b643) is CLIENT_CONNECTED
```
Everything works perfectly until you restart the HZ server service.
Note that on Docker, with the start/stop of a container, the IP address changes.
The client then looses the connection to the server and tries to reconnect.
But at this point the client will try to reconnect keeping the old IP address it found at startup time (although the JVM dns cache is set to 0).
In the logs, while the server is being restarted, I can see :
```
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.hz.client_1.IO.thread-in-0] com.hazelcast.client.connection.nio.ClientConnection.null hz.client_1 [hazelcast] [3.12.7] ClientConnection{alive=false, connectionId=1, channel=NioChannel{/172.20.0.40:57827->hazelcast/172.20.0.39:5701}, remoteEndpoint=[hazelcast]:5701, lastReadTime=2020-06-08 12:25:59.755, lastWriteTime=2020-06-08 12:25:54.447, closedTime=2020-06-08 12:25:59.755, connected server version=3.12.7} closed. Reason: Connection closed by the other side
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.hz.client_1.IO.thread-in-0] com.hazelcast.client.connection.ClientConnectionManager.null hz.client_1 [hazelcast] [3.12.7] Removed connection to endpoint: [hazelcast]:5701, connection: ClientConnection{alive=false, connectionId=1, channel=NioChannel{/172.20.0.40:57827->hazelcast/172.20.0.39:5701}, remoteEndpoint=[hazelcast]:5701, lastReadTime=2020-06-08 12:25:59.755, lastWriteTime=2020-06-08 12:25:54.447, closedTime=2020-06-08 12:25:59.755, connected server version=3.12.7}
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.client_1.cluster-] com.hazelcast.core.LifecycleService.null hz.client_1 [hazelcast] [3.12.7] HazelcastClient 3.12.7 (20200430 - bf4b643) is CLIENT_DISCONNECTED
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to cluster with name: hazelcast
hz-test-client              | 2020-06-08 12:25:59 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:25:59 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: Connection refused to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:25:59 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 1 of 2147483647.
hz-test-client              | 2020-06-08 12:26:05 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:10 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:10 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 2 of 2147483647.
hz-test-client              | 2020-06-08 12:26:15 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:20 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:20 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 3 of 2147483647.
hz-test-client              | 2020-06-08 12:26:25 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:30 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:30 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 4 of 2147483647.
hz-test-client              | 2020-06-08 12:26:35 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:40 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketTimeoutException
hz-test-client              | 2020-06-08 12:26:40 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 5 of 2147483647.
hz-test-client              | 2020-06-08 12:26:45 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:49 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:26:49 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 6 of 2147483647.
hz-test-client              | 2020-06-08 12:26:54 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:26:57 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:26:57 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 7 of 2147483647.
hz-test-client              | 2020-06-08 12:27:03 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:27:06 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:27:06 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 8 of 2147483647.
hz-test-client              | 2020-06-08 12:27:11 INFO [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Trying to connect to [hazelcast]:5701 as owner member
hz-test-client              | 2020-06-08 12:27:14 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Exception during initial connection to [hazelcast]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: No route to host to address hazelcast/172.20.0.39:5701
hz-test-client              | 2020-06-08 12:27:14 WARNING [hz.client_1.cluster-] com.hazelcast.client.connection.nio.ClusterConnectorService.null hz.client_1 [hazelcast] [3.12.7] Unable to get live cluster connection, retry in 5000 ms, attempt 9 of 2147483647.
```
but 172.20.0.39 is the old IP address of the hazelcast server, it has changed in docker since the restart and the client never attempts to resolve "hazelcast" again, leading to impossible reconnection.
It seems that when the client is executing its reconnection mechanism, it reuses the same connection object with the IP address resolved at startup time, but if it has changed since, it will never attempt to resolve the member address defined in the configuration again to reconnect.

I also tried the same thing with a 2 member cluster and it leads to the same errors.
On the server side whenever a member IP address changes, other members are capable of reconnecting to it automatically and see the address change, the issue is only present on the Java client side.
Tried with/without smart-routing but doesn't affect client behavior on this.
Tried with "hazelcast.diagnostics.enabled" on client side but didn't see anything more that would point me to a configuration issue or something onto which I can do anything.

Is this the expected behavior? 
Thanks for the feedback!

Hi, just a FYI, after [implementing the proposed solution on the tomcat replication session HZ manager](https://github.com/hazelcast/hazelcast-tomcat-sessionmanager/pull/92) (which is to listen for client shutdown and recreate a new one when necessary), I'm still experiencing some weird behaviors where client thinks it's still connected to a member whereas the member has already restarted and changed its IP address.
I must say this seems to happen only in race conditions where I restart the members multiple times in random ways and orders (I have a 2 member cluster in my test case).

In such case I'm getting errors like this one:
```
hz-test-client              | com.hazelcast.core.OperationTimeoutException: ClientInvocation{clientMessage = ClientMessage{connection=null, length=25597, operation=Map.Set, isRetryable=false, correlationId=1359, messageType=10f00, isEvent=false, isfragmented=false}, objectName = hz-test-client, target = partition 88, sendConnection = null} timed out because exception occurred after client invocation timeout 120000 ms. Current time: 2020-07-15 18:23:03.565. Start time: 2020-07-15 18:21:03.407. Total elapsed time: 120158 ms.
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.notifyException(ClientInvocation.java:264)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:173)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.retry(ClientInvocation.java:193)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.run(ClientInvocation.java:179)
hz-test-client              |      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
hz-test-client              |      at java.util.concurrent.FutureTask.run(FutureTask.java:266)
hz-test-client              |      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
hz-test-client              |      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
hz-test-client              |      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
hz-test-client              |      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
hz-test-client              |      at java.lang.Thread.run(Thread.java:748)
hz-test-client              |      at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
hz-test-client              |      at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
hz-test-client              | Caused by: java.io.IOException: No available connection to member 63b91b70-6eaf-4d8b-9833-83a6550822d5
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.SmartClientInvocationService.getConnection(SmartClientInvocationService.java:125)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.SmartClientInvocationService.invokeOnTarget(SmartClientInvocationService.java:118)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.SmartClientInvocationService.invokeOnPartitionOwner(SmartClientInvocationService.java:95)
hz-test-client              |      at com.hazelcast.client.impl.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:166)
hz-test-client              |      ... 11 more
```

In this case the client doesn't seem to somehow "mark" the connection as bad/lost (so that it falls in shutdown state, which would be then detected by the code I've added and would be restarted), the only way I'm able to get it back to life is to either restart the client application (the tomcat server) or restart the full HZ cluster (restart all the members, 2 actually in my test case).

Otherwise the solution seem to fix 99% of the issue and the tomcat replication plugin is now able to automatically reconnect whenever cluster members gets restarted, there seem to be only some few rare cases where it still falls in the above case. I'm trying to get a 100% reproducable scenario meanwhile.

--- a/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java
@@ -56,7 +56,6 @@ import com.hazelcast.internal.nio.ConnectionListener;
 import com.hazelcast.internal.nio.ConnectionType;
 import com.hazelcast.internal.serialization.InternalSerializationService;
 import com.hazelcast.internal.util.AddressUtil;
-import com.hazelcast.internal.util.ConcurrencyUtil;
 import com.hazelcast.internal.util.EmptyStatement;
 import com.hazelcast.internal.util.RuntimeAvailableProcessors;
 import com.hazelcast.internal.util.UuidUtil;
@@ -76,7 +75,6 @@ import java.io.IOException;
 import java.net.InetSocketAddress;
 import java.net.Socket;
 import java.net.SocketAddress;
-import java.net.UnknownHostException;
 import java.nio.channels.SocketChannel;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -127,7 +125,6 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
     private final ILogger logger;
     private final int connectionTimeoutMillis;
     private final HazelcastClientInstanceImpl client;
-    private final ConcurrentMap<Address, InetSocketAddress> inetSocketAddressCache = new ConcurrentHashMap<>();
     private final Collection<ConnectionListener> connectionListeners = new CopyOnWriteArrayList<>();
     private final NioNetworking networking;
     private final HeartbeatManager heartbeat;
@@ -157,6 +154,7 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
     private volatile UUID clusterId;
     private volatile ClientState clientState = ClientState.INITIAL;
     private volatile boolean connectToClusterTaskSubmitted;
+    private volatile boolean switchingToNextCluster;
 
     private enum ClientState {
         /**
@@ -200,13 +198,12 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
         this.authenticationTimeout = heartbeat.getHeartbeatTimeout();
         this.failoverConfigProvided = client.getFailoverConfig() != null;
         this.executor = createExecutorService();
-        this.shuffleMemberList = client.getProperties().getBoolean(SHUFFLE_MEMBER_LIST);
         this.clusterDiscoveryService = client.getClusterDiscoveryService();
-        this.isSmartRoutingEnabled = client.getClientConfig().getNetworkConfig().isSmartRouting();
         this.waitStrategy = initializeWaitStrategy(client.getClientConfig());
-        ClientConnectionStrategyConfig connectionStrategyConfig = client.getClientConfig().getConnectionStrategyConfig();
-        this.asyncStart = connectionStrategyConfig.isAsyncStart();
-        this.reconnectMode = connectionStrategyConfig.getReconnectMode();
+        this.shuffleMemberList = client.getProperties().getBoolean(SHUFFLE_MEMBER_LIST);
+        this.isSmartRoutingEnabled = client.getClientConfig().getNetworkConfig().isSmartRouting();
+        this.asyncStart = client.getClientConfig().getConnectionStrategyConfig().isAsyncStart();
+        this.reconnectMode = client.getClientConfig().getConnectionStrategyConfig().getReconnectMode();
     }
 
     private int initConnectionTimeoutMillis() {
@@ -414,7 +411,7 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
         ((ClientLoggingService) client.getLoggingService()).updateClusterName(nextContext.getClusterName());
 
         logger.info("Trying to connect to next cluster: " + nextContext.getClusterName());
-
+        switchingToNextCluster = true;
         if (doConnectToCandidateCluster(nextContext)) {
             client.waitForInitialMembershipEvents();
             fireLifecycleEvent(CLIENT_CHANGED_CLUSTER);
@@ -661,7 +658,9 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
 
             Channel channel = networking.register(currentClusterContext.getChannelInitializer(), socketChannel, true);
             channel.attributeMap().put(Address.class, target);
-            channel.connect(resolveAddress(target), connectionTimeoutMillis);
+
+            InetSocketAddress inetSocketAddress = new InetSocketAddress(target.getInetAddress(), target.getPort());
+            channel.connect(inetSocketAddress, connectionTimeoutMillis);
 
             TcpClientConnection connection = new TcpClientConnection(client, connectionIdGen.incrementAndGet(), channel);
 
@@ -757,8 +756,6 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
         for (TcpClientConnection activeConnection : activeConnections.values()) {
             activeConnection.close(null, new TargetDisconnectedException("Closing since client is switching cluster"));
         }
-
-        inetSocketAddressCache.clear();
     }
 
     @Override
@@ -876,18 +873,18 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
                 logger.fine("Checking the cluster: " + newClusterId + ", current cluster: " + this.clusterId);
             }
 
-            boolean initialConnection = activeConnections.isEmpty();
-            boolean changedCluster = initialConnection && this.clusterId != null && !newClusterId.equals(this.clusterId);
-            if (changedCluster) {
+            boolean clusterIdChanged = this.clusterId != null && !newClusterId.equals(this.clusterId);
+            if (clusterIdChanged) {
+                checkClientStateOnClusterIdChange(connection);
                 logger.warning("Switching from current cluster: " + this.clusterId + " to new cluster: " + newClusterId);
                 client.onClusterRestart();
             }
 
+            boolean connectionsEmpty = activeConnections.isEmpty();
             activeConnections.put(response.memberUuid, connection);
-
-            if (initialConnection) {
+            if (connectionsEmpty) {
                 clusterId = newClusterId;
-                if (changedCluster) {
+                if (clusterIdChanged) {
                     clientState = ClientState.CONNECTED_TO_CLUSTER;
                     executor.execute(() -> initializeClientOnCluster(newClusterId));
                 } else {
@@ -912,6 +909,30 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
             onConnectionClose(connection);
         }
         return connection;
+    }
+
+    private void checkClientStateOnClusterIdChange(TcpClientConnection connection) {
+        if (activeConnections.isEmpty()) {
+            //We only have single connection established
+            if (failoverConfigProvided) {
+                //If failover is provided, and this single connection is established after failover logic kicks in
+                // (checked via `switchingToNextCluster`), then it is OK to continue. Otherwise, we force the failover logic
+                // to be used by throwing `ClientNotAllowedInClusterException`
+                if (switchingToNextCluster) {
+                    switchingToNextCluster = false;
+                } else {
+                    String reason = "Force to hard cluster switch";
+                    connection.close(reason, null);
+                    throw new ClientNotAllowedInClusterException(reason);
+                }
+            }
+        } else {
+            //If there are other connections that means we have a connection to wrong cluster.
+            //We should not stay connected.
+            String reason = "Connection does not belong to this cluster";
+            connection.close(reason, null);
+            throw new IllegalStateException(reason);
+        }
     }
 
     private ClientMessage encodeAuthenticationRequest(Address toAddress) {
@@ -954,16 +975,6 @@ public class TcpClientConnectionManager implements ClientConnectionManager {
                     + "Expected partition count: " + partitionService.getPartitionCount()
                     + ", Member partition count: " + newPartitionCount);
         }
-    }
-
-    private InetSocketAddress resolveAddress(Address target) {
-        return ConcurrencyUtil.getOrPutIfAbsent(inetSocketAddressCache, target, arg -> {
-            try {
-                return new InetSocketAddress(target.getInetAddress(), target.getPort());
-            } catch (UnknownHostException e) {
-                throw rethrow(e);
-            }
-        });
     }
 
     private void initializeClientOnCluster(UUID targetClusterId) {

